## Assignment 4  Question 3

### 3a) The words that I have remove are "I", "we", "you", "he","she", "it", "they", "this","that","these","those","am","is","are","was","were","a","an","the","and","but","if","or","of","at","by","for". These words are stopwords that are not important in sentiment analysis. 

### 3b) Import the necessary libraries and modules

In [1]:
import turicreate as tc

### 3c) The words that I have remove are "I", "we", "you", "he","she", "it", "they", "this","that","these","those","am","is","are","was","were","a","an","the","and","but","if","or","of","at","by","for". These words are stopwords, they occur in abundance and they are usually not important or provide little informarion that can be used in sentiment analysis. We want to remove these words as much as we can to give the model more focus to the important information for prediction. 

In [2]:
movies = tc.SFrame('./IMDB_Dataset_mod.csv')
movies

Finished parsing file /home/qyu2/SupervisedML/Assignment4/IMDB_Dataset_mod.csv

Parsing completed. Parsed 100 lines in 0.387153 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 47401 lines. Lines per second: 65595.2

Finished parsing file /home/qyu2/SupervisedML/Assignment4/IMDB_Dataset_mod.csv

Parsing completed. Parsed 50000 lines in 0.751625 secs.

Columns:
	review	str
	sentiment	str

Rows: 50000

Data:
+-------------------------------+-----------+
|             review            | sentiment |
+-------------------------------+-----------+
| One other reviewers has me... |  positive |
| wonderful little productio... |  positive |
| thought wonderful way to s... |  positive |
| Basically there's family w... |  negative |
| Petter Mattei's "Love in T... |  positive |
| Probably my all-time favor... |  positive |
| sure would like to see res... |  positive |
| show amazing, fresh & inno... |  negative |
| Encouraged positive commen... |  negative |
| like original gut wrenchin... |  positive |
+-------------------------------+-----------+
[50000 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### 3d) 

In [3]:
movies['words'] = tc.text_analytics.count_words(movies['review'])
movies

Columns:
	review	str
	sentiment	str
	words	dict

Rows: 50000

Data:
+-------------------------------+-----------+-------------------------------+
|             review            | sentiment |             words             |
+-------------------------------+-----------+-------------------------------+
| One other reviewers has me... |  positive | {'darker': 1.0, 'touch': 1... |
| wonderful little productio... |  positive | {'done': 1.0, 'guided': 1.... |
| thought wonderful way to s... |  positive | {'go': 1.0, 'superman': 1.... |
| Basically there's family w... |  negative | {'them': 1.0, 'ignore': 1.... |
| Petter Mattei's "Love in T... |  positive | {'work': 1.0, 'anxiously':... |
| Probably my all-time favor... |  positive | {'dozen': 1.0, 'be': 1.0, ... |
| sure would like to see res... |  positive | {'it': 1.0, 'do': 1.0, 'th... |
| show amazing, fresh & inno... |  negative | {'awful': 1.0, 'just': 1.0... |
| Encouraged positive commen... |  negative | {'obsessives': 1.0, 'effor... |
| like original gut wrenchin... |  positive | {'br': 2.0, 'mom': 1.0, 'm... |
+-------------------------------+-----------+-------------------------------+
[50000 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### 3e) Randomly split the data into training and testing sets, with 80% of the data in the training set.

In [4]:
movies_train, movies_test = movies.random_split(0.8, seed=0)
#print(len(movies_train),len(movies_test))

### 3f) L1 penalty: 0.1 & L2 penalty: 0.1. 

In [5]:
model = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0.1, l2_penalty=0.1)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92686

Number of coefficients      : 92687

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 1.060027     | 0.500552          | 0.507493            |

| 1         | 2        | 0.000008  | 2.694162     | 0.925191          | 0.870130            |

| 2         | 3        | 0.000008  | 2.831999     | 0.956534          | 0.893606            |

| 3         | 4        | 0.000003  | 3.064562     | 0.970076          | 0.895105            |

| 4         | 5        | 0.000002  | 3.252782     | 0.974389          | 0.906094            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

### 3f) The final training accuracy is 0.974389 and the validation accuracy is 0.906094. 

In [6]:
results = model.evaluate(movies_test)
results

{'accuracy': 0.8908507223113965,
 'auc': 0.9513737692721433,
 'confusion_matrix': Columns:
 	target_label	str
 	predicted_label	str
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |   negative   |     positive    |  533  |
 |   positive   |     negative    |  555  |
 |   positive   |     positive    |  4393 |
 |   negative   |     negative    |  4487 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.8898116264938223,
 'log_loss': 0.36487880213914603,
 'precision': 0.8917986195696306,
 'recall': 0.8878334680679062,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +-----------+--------------------+-----+------+------+
 | threshold |        fpr         | tpr |  p   |  n   |
 +-----------+--------------------+-----+------+------+
 |    0.0    |        1.0         | 1.0 | 4948 | 5020 |
 

In [7]:
results['accuracy']

0.8908507223113965

### 3g) The testing accuracy is 0.890851. 

In [8]:
model1 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0.01, l2_penalty=0.1,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92554

Number of coefficients      : 92555

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.046768     | 0.499395          | 0.529471            |

| 1         | 2        | 0.000008  | 1.377478     | 0.923718          | 0.871628            |

| 2         | 3        | 0.000008  | 1.499448     | 0.962372          | 0.901099            |

| 3         | 4        | 0.000008  | 1.624564     | 0.974888          | 0.909590            |

| 4         | 5        | 0.000008  | 1.749176     | 0.977413          | 0.903596            |

| 5         | 6        | 0.000003  | 1.954738     | 0.982251          | 0.912587            |

| 6         | 7        | 0.000003  | 2.080116     | 0.983934          | 0.910090            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [9]:
model1_results = model1.evaluate(movies_test)
model1_results['accuracy']

0.8953651685393258

In [10]:
model2 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0.1, l2_penalty=0.01,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92335

Number of coefficients      : 92336

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.049303     | 0.500263          | 0.512987            |

| 1         | 2        | 0.000008  | 1.395765     | 0.925638          | 0.857642            |

| 2         | 3        | 0.000008  | 1.517913     | 0.958059          | 0.882118            |

| 3         | 4        | 0.000003  | 1.722725     | 0.970523          | 0.884116            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [11]:
model2_results = model2.evaluate(movies_test)
model2_results['accuracy']

0.8865369181380417

In [12]:
model3 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0.01, l2_penalty=0.01,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92511

Number of coefficients      : 92512

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.077801     | 0.501078          | 0.497502            |

| 1         | 2        | 0.000008  | 1.402203     | 0.925454          | 0.864136            |

| 2         | 3        | 0.000008  | 1.524536     | 0.956219          | 0.882118            |

| 3         | 4        | 0.000003  | 1.728599     | 0.969261          | 0.884615            |

| 4         | 5        | 0.000002  | 1.888555     | 0.974967          | 0.898102            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [13]:
model3_results = model3.evaluate(movies_test)
model3_results['accuracy']

0.8910513643659711

In [14]:
model4 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0, l2_penalty=0.01,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92605

Number of coefficients      : 92606

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 0.098925     | 0.926058          | 0.858641            |

| 1         | 4        | 1.000000  | 0.223141     | 0.974283          | 0.883616            |

| 2         | 5        | 1.000000  | 0.303435     | 0.986695          | 0.891608            |

| 3         | 6        | 1.000000  | 0.385533     | 0.986064          | 0.858142            |

| 4         | 8        | 0.750248  | 0.508286     | 0.994084          | 0.885614            |

| 9         | 15       | 1.000000  | 1.007532     | 0.999527          | 0.850649            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [15]:
model4_results = model4.evaluate(movies_test)
model4_results['accuracy']

0.8685794542536116

In [16]:
model5 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0.01, l2_penalty=0,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92697

Number of coefficients      : 92698

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.045684     | 0.500868          | 0.501499            |

| 1         | 2        | 0.000008  | 1.374778     | 0.924428          | 0.854645            |

| 2         | 3        | 0.000008  | 1.505886     | 0.957928          | 0.879620            |

| 3         | 4        | 0.000003  | 1.710863     | 0.970366          | 0.895604            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [17]:
model5_results = model5.evaluate(movies_test)
model5_results['accuracy']

0.8855337078651685

In [18]:
model6 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0, l2_penalty=0.1,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92550

Number of coefficients      : 92551

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 0.100475     | 0.926374          | 0.852148            |

| 1         | 4        | 1.000000  | 0.222575     | 0.974520          | 0.879620            |

| 2         | 5        | 1.000000  | 0.303677     | 0.986484          | 0.882118            |

| 3         | 6        | 1.000000  | 0.386979     | 0.985617          | 0.848651            |

| 4         | 8        | 0.728244  | 0.509772     | 0.994005          | 0.877123            |

| 9         | 15       | 1.000000  | 1.017037     | 0.999606          | 0.865634            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [19]:
model6_results = model6.evaluate(movies_test)
model6_results['accuracy']

0.8703852327447833

In [20]:
model7 = tc.logistic_classifier.create(movies_train, features=['words'], target='sentiment', l1_penalty=0.1, l2_penalty=0,)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92637

Number of coefficients      : 92638

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 0.050077     | 0.500473          | 0.508991            |

| 1         | 2        | 0.000008  | 1.368846     | 0.925454          | 0.864136            |

| 2         | 3        | 0.000008  | 1.491612     | 0.958112          | 0.881618            |

| 3         | 4        | 0.000003  | 1.699047     | 0.970234          | 0.891109            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [21]:
model7_results = model7.evaluate(movies_test)
model7_results['accuracy']

0.8841292134831461

### 3h)
| Model | L1 Penalty | L2 Penalty | Training Accuracy | Validation Accuracy | Testing Accuracy |
| :-: | :-: | :-: | :-: | :-: |:-: |
| model | 0.1 | 0.1 | 0.974389 | 0.906094 | 0.890851 |
| model1 | 0.01 | 0.1 | 0.983934 | 0.910090 | 0.895365|
| model2 | 0.1 | 0.01 | 0.970523 | 0.884116 | 0.886537 |
| model3 | 0.01 | 0.01 | 0.974967 | 0.898102 | 0.891051 |
| model4 | 0 | 0.01 | 0.999527 | 0.850649 | 0.868579 |
| model5 | 0.01 | 0 | 0.970366 | 0.895604 | 0.885534 |
| model6 | 0 | 0.1 | 0.999606 | 0.865634 | 0.870385 |
| model7 | 0.1 | 0 | 0.970234 | 0.891109 | 0.884129 |

### 3i) When we train a model, we want to avoid overfitting. From the table above, model 4 and model 6 have the highest training accuracies of 99.96%. These two models did well on the training; however, the testing accuracies are both low compared with other models. Therefore, model 4 and model 6 might have a little bit of overfitting. The first model, model 1, and model 3 are the only three models that have testing accuracy over 0.89. The difference between training accuracy and testing accuracy for these three models is 0.08, which is lower than the difference from other models. A validation dataset is used to check how well the model is able to make predictions based on data it hasn't seen before. Model 1 with a L1 penalty of 0.01 and a L2 penalty of 0.1 seen to doing well on both validation and testing. Model 1 has the highest validation accuracy of 0.91 and the highest testing accuracy of 0.90. Model 1 is able to perform with 91% accuracy on new data. The training accuracy is 0.98, which is good. Therefore, Model 1 is the optimal choice of L1 and L2.  